### PR TITLE
Add null checks in DruidSchema

### PR DESCRIPTION
**Pull request by surekhasaharan**
Fix for #6826

### CODE REVIEW

**Comment by surekhasaharan** [comment]
@egor-ryashin I think theoretically, this issue can happen, it'd require some effort to create a unit test for multithreaded case, let me think about how can I add such test.

**Comment by surekhasaharan** [comment]
> Could you create a unit test which checks if the issue exists?

@egor-ryashin added unit tests.

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java [hunk 1]
```diff
@@ -402,7 +402,7 @@ private void addSegment(final DruidServerMetadata server, final DataSegment segm
     }
   }
 
-  private void removeSegment(final DataSegment segment)
+  protected void removeSegment(final DataSegment segment)
```
Comment (jihoonson) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java]: Please add `@VisibleForTesting` to newly exposed methods.

Comment (surekhasaharan) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java]: Yes, forgot to add. Will do.

File: sql/src/test/java/org/apache/druid/sql/calcite/schema/DruidSchemaTest.java [hunk 2]
```diff
@@ -237,4 +238,36 @@ public void testGetTableMapFoo2()
...
+  @Test
+  public void testNullDatasource() throws IOException
+  {
+    Map<DataSegment, SegmentMetadataHolder> segmentsMetadata = schema.getSegmentMetadata();
+    Set<DataSegment> segments = segmentsMetadata.keySet();
+    Assert.assertEquals(segments.size(), 3);
+    final DataSegment segmentToRemove = segments.stream().findFirst().orElse(null);
+    Assert.assertFalse(segmentToRemove == null);
+    schema.removeSegment(segmentToRemove);
+    schema.refreshSegments(segments); // can cause NPE without dataSourceSegments null check in DruidSchema#refreshSegmentsForDataSource
```
Comment (egor-ryashin) [inline sql/src/test/java/org/apache/druid/sql/calcite/schema/DruidSchemaTest.java]: It's unclear that we remove the only segment of 'foo2' datasource (and not a segment from `foo` datasource) as `getSegmentMetadata` returned Map doesn't guarantee order. Which means it's unclear if we test for `dataSourceSegments == null`

File: sql/src/test/java/org/apache/druid/sql/calcite/schema/DruidSchemaTest.java [hunk 3]
```diff
@@ -237,4 +238,36 @@ public void testGetTableMapFoo2()
     Assert.assertEquals("m1", fields.get(2).getName());
     Assert.assertEquals(SqlTypeName.BIGINT, fields.get(2).getType().getSqlTypeName());
   }
+
+  @Test
+  public void testNullDatasource() throws IOException
+  {
+    Map<DataSegment, SegmentMetadataHolder> segmentsMetadata = schema.getSegmentMetadata();
```
[the remark this answers is not in the document]
Comment (surekhasaharan) [reply sql/src/test/java/org/apache/druid/sql/calcite/schema/DruidSchemaTest.java]: sure, renamed

File: sql/src/test/java/org/apache/druid/sql/calcite/schema/DruidSchemaTest.java (hunk 2 above)
Comment (surekhasaharan) [reply sql/src/test/java/org/apache/druid/sql/calcite/schema/DruidSchemaTest.java]: yeah, changed to get segment with datasource 'foo2'

**Pull request merged**